### PR TITLE
Add test for init_process_group timeout

### PR DESCRIPTION
**Pull request by wconstab**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #112893
* __->__ #112803
* #112611

### CODE REVIEW

File: test/distributed/test_c10d_nccl.py [hunk 1]
```diff
@@ -1199,6 +1200,48 @@ def test_close_pg(self):
...
+        opts = dist.ProcessGroupNCCL.Options()
+        opts._timeout = timedelta(seconds=123)
+        with warnings.catch_warnings(record=True) as w:
+            dist.init_process_group(**base_opts, pg_options=opts)
+            # TODO(whc) i verified that we are indeed emitting this warning, and i can't figure out why i can't catch it.
+            # self.assertEqual(len(w), 1)
+            # self.assertTrue("pg_options._timeout was specified" in str(w[-1].message))
+        _check_nccl_timeout(torch.distributed.distributed_c10d.default_pg_timeout)
+        dist.destroy_process_group()
+
```
Comment (H-Huang) [inline test/distributed/test_c10d_nccl.py:1243]: nit: could have 1 more example doing `init_process_group` using both options and timeout args, as pyper does

**Review by H-Huang — APPROVED** [review]
Looks great! Thanks for adding

File: test/distributed/test_c10d_nccl.py (hunk 1 above)
Comment (wconstab) [reply test/distributed/test_c10d_nccl.py:1243]: added

**Pull request closed without merging**